### PR TITLE
basic keystore implementation

### PR DESCRIPTION
**Pull request by whyrusleeping**
Thank you to @kyledrake for prompting me to just write this already

License: MIT
Signed-off-by: Jeromy <why@ipfs.io>

### CODE REVIEW

**Comment by haadcode** [comment]
Oooh! Excellent stuff! Big fan of your work J :)

Thank you @kyledrake for prompting @whyrusleeping to just write this already.

File: keystore/keystore.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+	b, err := k.Bytes()
+	if err != nil {
+		return err
+	}
+
+	kp := filepath.Join(ks.dir, name)
+
+	_, err = os.Stat(kp)
+	if err == nil {
```
Comment (Kubuxu) [inline keystore/keystore.go]: To be safe I would use `os.IsExist` and change it to `ErrKeyExists` then, else passthrough the error.

File: keystore/keystore.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+		return ErrKeyExists
+	}
+
+	fi, err := os.Create(kp)
+	if err != nil {
+		return err
+	}
+	defer fi.Close()
+
+	_, err = fi.Write(b)
```
Comment (Kubuxu) [inline keystore/keystore.go]: I thought you wanted to use multicodec so we can migrate much easier?

**Comment by Kubuxu** [comment]
Also there are absolutely no tests for it.

**Comment by whyrusleeping** [comment]
> Also there are absolutely no tests for it.

Yeah yeah, i'm getting there.

File: keystore/keystore.go (hunk 1 above)
Comment (whyrusleeping) [inline keystore/keystore.go]: How would you use `os.IsExist` here?

Comment (Kubuxu) [inline keystore/keystore.go]: sorry, IsNotExist

Comment (whyrusleeping) [inline keystore/keystore.go]: ah! okay, i see what you want

**Comment by wigy-opensource-developer** [comment]
@whyrusleeping I love the new feature, thanks for implementing it.

What is missing is the renewal of the IPNS records every 4 hours. The namesys/republisher uses a so-called keybook in the peerstore that only contains the `self` key. Probably the republisher needs to use the keystore, too, to enumerate all IPNS records that need a renewal.

As I currently understand, the republisher/reprovider is the only part of the network that fixes data loss caused by peers going offline or leaving the network. If there is no automatic recovery from it in the DHT layer, in the current state of the code, the `ipns name publish --key notself /ipfs/QmBlah` needs to be executed regularly as a workaround to make sure the name can be resolved.

**Comment by whyrusleeping** [comment]
Yeah, that's the next part that needs to be done. Can you open a new issue
for discussing how we should go about doing that? Configuration UX will be
important

**Comment by Kubuxu** [comment]
I have opened it already: https://github.com/ipfs/go-ipfs/issues/3537

**Pull request merged**